xl2times/datatypes.py

```diff
@@ -161,14 +160,38 @@ class TimesModel:
     time_periods: DataFrame = field(default_factory=DataFrame)
     units: DataFrame = field(default_factory=DataFrame)
     start_year: int = field(default_factory=int)
-    data_years: Tuple[int] = field(default_factory=tuple)
-    model_years: Tuple[int] = field(default_factory=tuple)
-    past_years: Tuple[int] = field(default_factory=tuple)
     files: Set[str] = field(default_factory=set)
 
+    @property
     def external_regions(self) -> Set[str]:
         return self.all_regions.difference(self.internal_regions)
 
+    @property
+    def data_years(self) -> Set[int]:
+        """
+        data_years are years for which there is data specified.
+        """
+        data_years = set()
+        for attributes in [self.attributes, self.uc_attributes]:
+            if not attributes.empty:
+                data_years.update(attributes["year"].astype(int).values)
+        # Remove interpolation rules before return
+        return {y for y in data_years if y >= 1000}
```

Comment:
These changes mean that `data_years` etc are recomupted every time they are called as `model.data_years`. Did we want that behaviour for something in particular? Performance wise the new version could be slower, especially if `attributes` is a large table.

Reply:
Currently we don't use `model.data_years` for any processing. It is only needed to generate a corresponding set for GAMS. I expect calling it once after all the processing is done should suffice. I guess we could cache the value?